### PR TITLE
CI: Add ccache to GitHub Actions build tests

### PR DESCRIPTION
**Pull request by kinkie**
Depending on an individual build test, adding ccache currently reduces
build test time 35-60% (e.g., from 6m42s to 4m18s and from 17m to 7m).

Problematic ccache entries can be removed using GitHub API. For example:

    gh -R squid-cache/squid cache list
    gh -R squid-cache/squid cache delete <cache-key|"all">

Do not add ccache to MacOS build tests because I had mixed experience
with that optimization on that OS.

### CODE REVIEW

**Comment by kinkie** [comment]
A previous attempt failed with a permission problem in the squid-cache repository
but not in my own. Creating a new PR to debug it

**Comment by kinkie** [comment]
This is now working, the downside is that I had to fork the ccache-action repository into the squid-cache project
to authorize it.

**Comment by kinkie** [comment]
Example of a run on a cold cache: https://github.com/kinkie/squid/actions/runs/11641544312/job/32420173615 . Single job run duration: 13-14 minutes. 
On a warm cache: https://github.com/kinkie/squid/actions/runs/11645646184/job/32428911706 . Single job duration: 5-6 minutes

File: .github/workflows/quick.yaml [hunk 1]
```diff
@@ -127,11 +127,18 @@ jobs:
...
 
       - name: Checkout sources
         uses: actions/checkout@v4
 
+      - name: Setup ccache
+        uses: squid-cache/ccache-action@v1.2.14
+        with:
+          verbose: 2 # default 0
+          append-timestamp: false
+          key: ${{ github.job }}-${{ matrix.os }}-${{ matrix.compiler.CC }}-${{ matrix.layer.nick }}
```
Comment (rousskov) [inline .github/workflows/quick.yaml:140]: We do not add github.job to log filenames, naturally assuming/implying that no two jobs would test the same os-compiler-layer combo. If possible, let's not add github.job to cache key either (to avoid sending the opposite/contradictory signal).

```suggestion
          key: ${{ matrix.os }}-${{ matrix.compiler.CC }}-${{ matrix.layer.nick }}
```

Same for ccache key in the other file, of course.

File: .github/workflows/quick.yaml [hunk 2]
```diff
@@ -127,11 +127,18 @@ jobs:
...
+
       - name: Run build on Linux
         run: ./test-builds.sh ${{ matrix.layer.name }}
```
Comment (rousskov) [inline .github/workflows/quick.yaml:143]: Why are we not adding ccache to all existing build tests, but not MacOS tests? Ccache Action [docs](https://github.com/squid-cache/ccache-action) mention that the Action "works" on MacOS.

File: .github/workflows/quick.yaml [hunk 3]
```diff
@@ -116,8 +116,8 @@ jobs:
     name: build-tests(${{ matrix.os }},${{ matrix.compiler.CC }},${{ matrix.layer.nick }})
 
     env:
-      CC: ${{ matrix.compiler.CC }}
-      CXX: ${{ matrix.compiler.CXX }}
+      CC: /usr/bin/ccache ${{ matrix.compiler.CC }}
+      CXX: /usr/bin/ccache ${{ matrix.compiler.CXX }}
```
Comment (rousskov) [inline .github/workflows/quick.yaml:120]: When things go wrong (e.g., bad/stale info gets cached as it has happened in the past on Jenkins IIRC), how to erase ccache (without changing our workflow files)? I could not quickly find instructions in Ccache Action docs, but there is probably a way to do that from the command line and/or GitHub interface?

**Review by rousskov — CHANGES_REQUESTED** [review]
FWIW, time savings per build job are clearly visible on GitHub Actions [usage page](https://github.com/squid-cache/squid/actions/runs/11650562408/usage?pr=1935) for this PR. I rerun all tests, so the first build test (with the same name) corresponds to cold cache and the second (with the same name, on the next line) corresponds to warm cache. I have updated PR description with that info.

I adjusted PR title to add "CI" prefix, to improve GitHub Actions capitalization consistency, and to specify that ccache is used for build tests.

Please adjust further as needed, of course.

File: .github/workflows/quick.yaml (hunk 2 above)
Comment (kinkie) [reply .github/workflows/quick.yaml:143]: My own experience with ccache is mixed, it hasn't worked well for me so I am not using it on that OS

File: .github/workflows/quick.yaml (hunk 3 above)
Comment (kinkie) [reply .github/workflows/quick.yaml:120]: `gh -R squid-cache/squid cache list`
and
`gh -R squid-cache/squid cache delete <cache-key|"all">`

can help

File: .github/workflows/quick.yaml (hunk 1 above)
Comment (kinkie) [reply .github/workflows/quick.yaml:140]: I do not see that implication. We store the log filenames as they are, and they are distinguished by the job they refer to. Furthermore, these jobs are write-most and rarely read.

The cache does not fulfill the same objective, it is an internal implementation detail which we want to keep separate to avoid thrashing it. The default cache size is 0.5gb, which is enough to keep about 3 run's worth of uncacheable content (e.g. content with a difference in squid.h or autoconf.h). Keeping it separate per job is a way to maximise its effectiveness, letting github's eviction policies take care of not overflowing it.

In other words, I'd leave github.job in here; if you feel that there is a risk of confusion (I don't, but I accept your opinion) we can add a comment.

File: .github/workflows/slow.yaml [hunk 4]
```diff
@@ -49,13 +49,20 @@ jobs:
...
 
     steps:
       - name: Checkout Sources
         uses: actions/checkout@v4
 
+      - name: Setup ccache
+        uses: squid-cache/ccache-action@v1.2.14
+        with:
+          verbose: 2 # default 0
+          append-timestamp: false
```
Comment (rousskov) [inline .github/workflows/slow.yaml:63]: Do we have to make append-timestamp false? I could not tell from ccache-action "foo is foo" documentation; ccache-action [examples](https://github.com/hendrikmuhs/ccache-action) do not use this customization, and presumably work even when append-timestamp is true (by default), so I wonder whether/why we need this line.

**Review by rousskov — CHANGES_REQUESTED** [review]
Great progress, thank you! Only one old [issue](https://github.com/squid-cache/squid/pull/1935/files#r1827129658) remains unresolved AFAICT. I also asked one new question.

File: .github/workflows/slow.yaml (hunk 4 above)
Comment (kinkie) [reply .github/workflows/slow.yaml:63]: In my understanding, we do or we will not be able to reuse ccache across builds. I can experiment tho

Comment (rousskov) [reply .github/workflows/slow.yaml:63]: > In my understanding, we do or we will not be able to reuse ccache across builds.

... because ccache-action timestamps change slowly enough to stay the same within one workflow build/execution, even if individual workflow jobs could run minutes apart from each other? That would be a really weird/dangerous design!

> I can experiment tho

Thank you. I think such an experiment is worth it if it may result in keeping this poorly understood customization away from our workflows.

Comment (kinkie) [reply .github/workflows/slow.yaml:63]: Apparently ccache-action enumerates caches, so it can pick one and resstore it; I suppose there is a key somewhere. I could look through the code but for now it seems it "just works".

I wouldn't call the customization as "poorly understood" but if we can stick with the default, all the better

File: .github/workflows/quick.yaml (hunk 1 above)
Comment (kinkie) [reply .github/workflows/quick.yaml:139]: The actual key used to restore the cache from a sample job is
`ccache-build-tests-ubuntu-22.04-gcc-default-2024-11-05T16:19:52.056Z`

The key question is: what is it namespaced to? The short answer is, it's hard to tell.
The code restoring the cache is at https://github.com/actions/cache/blob/main/src/saveImpl.ts , but it has so many dependencies into other workflows that it would take a fair amount of time to trace.

Sensible namespace scopes could be:
- the github account
- the repository
- the branch
- the workflow (file)
- the job

Documentation at https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows seems to imply (but never states) that the scope is the branch (with additional branches added to the scope when the action is evaluated for a PR)

I'd rather approach this the other way around: can anyone demonstrate that having the job name in the key will hurt, besides the confusion issue you mention which can be trivially solved by a comment?
Is the potential harm of creating confusion in a scope that's unlikely to ever be considered again by anyone else even worth the time we are spending in the discussion?

File: .github/workflows/slow.yaml (hunk 4 above)
Comment (rousskov) [reply .github/workflows/slow.yaml]: > Apparently ccache-action enumerates caches, so it can pick one and resstore it; I suppose there is a key somewhere.

Yes, we define that key (there is even another change request about that definition).

 
> I wouldn't call the customization as "poorly understood" 

I do not understand it. Documentation says pretty much nothing. Looking through code does not offer enough clues (for me). Cache entry creation time (or equivalent) is probably (or should be) available without appending that time to the entry name or necessitating adding a configuration option about it. Your understanding has not been enough to explain why this customization is needed (or how it works) to me. Nobody else is participating in developing this understanding. I think all these factors, collectively, qualify as "poorly understood" :-)!

> if we can stick with the default, all the better

Exactly!

File: .github/workflows/quick.yaml (hunk 1 above)
Comment (rousskov) [reply .github/workflows/quick.yaml:139]: > I'd rather approach this the other way around

I understand the positives behind the other way around, but have to insist that, given the current resources and responsibilities, we justify additions/changes (rather than adding/changing unless proven wrong).

> can anyone demonstrate that having the job name in the key will hurt, besides the confusion issue you mention which can be trivially solved by a comment? 

There is no confusion that I know of. Adding a comment does not solve the inconsistency (for the lack of a better word) problem that this change request has identified.

> Is the potential harm of creating confusion in a scope that's unlikely to ever be considered again by anyone else even worth the time we are spending in the discussion?

We are definitely wasting time here, but that waste started when the inconsistency in question was introduced, not when I dared to flag it. Accepting my recommendation that would address it would take a few minutes. If that recommendation turns out to be wrong, we will learn something useful about these workflows _and_ use that knowledge to improve them (as opposed to adding a comment explaining that inconsistencies are there "just in case" or "because this code works" or whatever that excuse is going to be).

> The key question is: what is it namespaced to?

GitHub [Cache Action](https://github.com/actions/cache/tree/main) namespace is a GitHub repository AFAICT. There is also a cache entry "versioning" mechanism that prevents (by default) cross-OS sharing of cache entries. That default protection can be disabled via enableCrossOsArchive.

FWIW, the above is _not_ the key question IMO. The key question is: Can two distinct os-compiler-layer builds co-exist across our current workflows for a given git commit? If the answer (to a meaningful interpretation of that question) is yes, then we need to fix our existing and PR workflows that assume the opposite answer! If the answer is no, then there is no need for this PR to add a job name to the existing, "working", and "reasonable" os-compiler-layer build key.

For example, if we were to test _both_ OpenSSL and GnuTLS builds for the same os-compiler-layer, then os-compiler-layer key would stop working.

File: .github/workflows/quick.yaml [hunk 5]
```diff
@@ -127,11 +127,17 @@ jobs:
...
           sudo apt-get --quiet=2 update
           sudo apt-get --quiet=2 build-dep squid
-          sudo apt-get --quiet=2 install linuxdoc-tools libtool-bin ${{ matrix.compiler.CC }}
+          sudo apt-get --quiet=2 install linuxdoc-tools libtool-bin ${{ matrix.compiler.CC }} ccache
 
       - name: Checkout sources
         uses: actions/checkout@v4
 
+      - name: Setup ccache
+        uses: squid-cache/ccache-action@v1.2.14
```
Comment (rousskov) [inline .github/workflows/quick.yaml:136]: @kinkie, if possible, please watch the upstream ccache-action repository for updates to reduce the chances that our new clone misses some useful ones. Thank you.

**Review by rousskov — APPROVED** [review]
Thank you for addressing my concerns. While I am worried about subtle problems this caching may introduce, I think the measured time savings are worth the risk in this case.

File: .github/workflows/quick.yaml (hunk 5 above)
Comment (kinkie) [reply .github/workflows/quick.yaml:136]: Yes. I'll check it from time to time

**Comment by rousskov** [comment]
> This is now working, the downside is that I had to fork the ccache-action repository into the squid-cache project
to authorize it.

```
squid-cache/ccache-action@v1.2.14 is not allowed to be used in measurement-factory/squid.
Actions in this workflow must be: within a repository owned by measurement-factory,
created by GitHub, or matching the following: [a list of explicitly allowed actions].
```

@kinkie, forking that well-known ccache repository forces others to clone Squid Project fork of ccache-action or allow our fork in their GitHub Actions configuration. I believe #1943 offers an overall better solution for the same problem because that solution does not rely on cloning well-known repositories.
